### ISSUE TITLE
Java: VirtualDispatch doesn't work properly with Default methods in Interfaces

### ISSUE THREAD

**Issue by atorralba** [github/codeql#6405]
With the introduction of `default` methods in Interfaces, some assumptions made in our `VirtualDispatch` library are no longer true (mainly, that an Interface method is always abstract and has no implementation). This causes flow analysis to fail when the flow goes through a `default` method.

## Example

Given a query that looks for viable implementations of a method access calling a method named `validate`:

~~~codeql
import java
import semmle.code.java.dispatch.VirtualDispatch

from VirtCalledSrcMethod top, SrcRefType tsrc, RefType t
where
  exists(MethodAccess ma | ma.getMethod().hasName("validate") |
    ma.getMethod().getSourceDeclaration() = top
  )
select viableMethodImpl(top, tsrc, t)
~~~

It correctly finds one result (`InputSource::validate`) in the following source code:

~~~java
public class RegularMethodsTest {

    public void input(InputSource source) {
        source.validate();
    }

    class InputSource {

        private String taintedField;

        public String getTaintedField() {
            return taintedField;
        }

        public void validate() {
            String tainted = getTaintedField();
            sink(tainted);
        }

        private void sink(String input) {}
    }
}
~~~~

But fails to return a result when analyzing the following source code:

~~~java
public class DefaultMethodsTest {

    public void input(InputSource source) {
        source.validate();
    }

    interface InputSource {

        public String getTaintedField();

        default void validate() {
            String tainted = getTaintedField();
            sink(tainted);
        }

        void sink(String input);
    }

}
~~~

**Comment by aschackmull** [comment]
Have you observed FN flow in real code due to this? I can reproduce with the example given above, but if I add an implementation of the interface that inherits the default method then dispatch should work
```
    class B implements InputSource {
        public String getTaintedField() { return ""; }
        public void sink(String input) { }
    }
```
This behaviour for `default` methods on interfaces is the same as for other non-abstract methods present on `abstract` classes. The dispatch library looks for an actual class that can be instantiated to exist somewhere in the type hierarchy that inherits the `default` method before concluding that the method is a possible dispatch target.

This digs into the open-/closed-world assumption trade-offs that are core to the calculation of dispatch. If you've observed this to be an issue in practice then we'll need to tweak that trade-off.

**Comment by atorralba** [comment]
I reported this because I observed lost flow in a synthetic application I was creating for a talk based on a Spring demo application. I've not seen this happen in the real world whatsoever (although spotting FNs is always difficult). Now that it seems this is an intended design decision and not a bug, I'm fine with not doing anything for the moment and revisit this if we see it happening in real-world applications.
